### ISSUE TITLE
doc.retokenize() merges all tokens together

### ISSUE THREAD

**Issue by bdewilde** [explosion/spaCy#2193]
I'd like to take advantage of the new `doc.retokenize()` context manager, as implemented in https://github.com/explosion/spaCy/pull/2172, but every token seems to get merged together rather than just the spans. To reproduce:

```
import spacy
from spacy.attrs import intify_attrs

en = spacy.load('en')
doc = en('Matthew Honnibal is the creator of spaCy.')
assert len(doc) == 8
ents = [(ent.start_char, ent.end_char, ent.label)
        for ent in doc.ents]
assert len(ents) == 2
with doc.retokenize() as retokenizer:
    string_store = doc.vocab.strings
    for start, end, label in ents:
        retokenizer.merge(
            doc[start: end],
            attrs=intify_attrs({'ent_type': label}, string_store))
assert len(doc) == 1  # this should not be true, right?
```

## Your Environment

* **spaCy version:** 2.0.11
* **Platform:** Darwin-17.4.0-x86_64-i386-64bit
* **Python version:** 3.6.5
* **Models:** en_core_web_sm, en

**Comment by ines** [comment]
I think I know what might be going wrong here: When you get your entity tuples, you're taking the `ent.start_char` and `ent.end_char`, which are the *character offsets*, not the token indices. Those are `ent.start` and `ent.end`. So when you're creating your span as `doc[start:end]`, you're actually getting the span of `doc[0:16]`, which is the entire sentence. (spaCy lets you do that and doesn't complain, because this is the default behaviour of slices in Python.)

TL;DR: Just tried it with `(ent.start, ent.end, ent.label)` and it looks like works as expected! 😊

**Comment by bdewilde** [comment]
Oh, gosh, you're totally right — thanks, Ines! I didn't see any docs for this new feature, so was going off the related `Doc.merge()` feature rather than checking directly in the source code. I got confused because `Doc.merge()` takes character indexes, while `Retokenizer.merge()` takes spans accessed via token indexes. I'm sure a brief usage example somewhere would make the difference clear. :)
